Move to Sqlite

bluekirby1111 opened the pull request.

randomcash committed:
Remove postgres, add sqlite

randomcash committed:
Update models

randomcash committed:
Update dependencies

randomcash committed:
Update models and types

mikonieminen commented on migrations/00000000000000_diesel_initial_setup/up.sql:
```diff
@@ -47,10 +47,10 @@ CREATE TABLE IF NOT EXISTS transactions (
+    validated INTEGER NOT NULL,
     bundle_id CHAR(43) REFERENCES bundle(id),
-    sent_to_leader bool NOT NULL,
+    sent_to_leader INTEGER NOT NULL,
```
Please use `BOOLEAN`, it's standard type. Even when it gets translated into integer, it's good indicator for the schema reader about the actual type of this field.

bluekirby1111 replied:
Sqlite doesn't have boolean type. Any suggestions on this?

https://stackoverflow.com/questions/843780/store-boolean-value-in-sqlite

mikonieminen commented on migrations/00000000000000_diesel_initial_setup/up.sql:
```diff
@@ -47,10 +47,10 @@ CREATE TABLE IF NOT EXISTS transactions (
     block_actual BIGINT,
-    signature bytea NOT NULL,
-    validated bool NOT NULL,
+    signature blob NOT NULL,
```
Write the field type in all caps as `BLOB`.

mikonieminen commented on migrations/00000000000000_diesel_initial_setup/up.sql:
```diff
@@ -47,10 +47,10 @@ CREATE TABLE IF NOT EXISTS transactions (
-    signature bytea NOT NULL,
-    validated bool NOT NULL,
+    signature blob NOT NULL,
+    validated INTEGER NOT NULL,
```
Please use BOOLEAN, it's standard type. Even when it gets translated into integer, it's good indicator for the schema reader about the actual type of this field.

mikonieminen commented on src/cron/arweave/arweave.rs:
```diff
@@ -48,8 +48,8 @@ pub struct TransactionData {
-    pub timestamp: i64,
-    pub height: i64,
+    pub timestamp: i32,
+    pub height: i32,
```
Why are these changed? Especially for the timestamp is too small.

bluekirby1111 replied:
Ok, I figured out how to use i64 with sqlite

mikonieminen commented on src/cron/arweave/arweave.rs:
```diff
@@ -115,7 +115,7 @@ pub struct Arweave {
 pub struct GqlVariables {
     pub owners: Vec<String>,
-    pub first: i64,
+    pub first: i32,
```
What is this field used for and why to change it from `i64` to `i32`?

bluekirby1111 replied:
Diesel implementations with Sqlite only supports i32

mikonieminen commented on src/cron/bundle.rs:
```diff
@@ -30,7 +30,7 @@ pub struct Bundler {
 #[derive(Serialize, Deserialize, Default, Debug)]
 pub struct TxReceipt {
-    block: i64,
+    block: i32,
```
`i32` feels a bit small for storing block number. Why is this changed? On the other hand, I would rather use unsigned integers here.

mikonieminen commented on src/cron/bundle.rs:
```diff
@@ -222,14 +222,14 @@ where
+                        validated: 1,
                         bundle_id: Some(bundle_tx.tx_id.clone()),
-                        sent_to_leader: false,
+                        sent_to_leader: 0,
```
I don't really like using `1` and `0` for boolean values, is there no way to use actual boolean values for these fields with SQLite?

mikonieminen commented on src/cron/leader.rs:
```diff
@@ -24,7 +24,7 @@ pub struct ValidatorSignature {
     id: String,
     signature: String,
-    block: i64,
+    block: i32,
```
Again, feels a bit small field type for block number.

mikonieminen commented on src/database/models.rs:
```diff
@@ -5,40 +5,40 @@ use serde::Serialize;
-    pub epoch: i64,
-    pub block_promised: i64,
-    pub block_actual: Option<i64>,
+    pub id: Option<String>,
```
Is this really optional?

mikonieminen commented on src/database/schema.rs:
```diff
@@ -1,34 +1,34 @@
-        id -> Bpchar,
-        owner_address -> Nullable<Bpchar>,
-        block_height -> Int8,
+        id -> Nullable<Text>,
```
Why is this column changed to allow `NULL`?

mikonieminen commented on src/database/schema.rs:
```diff
@@ -1,34 +1,34 @@
-        owner_address -> Nullable<Bpchar>,
-        block_height -> Int8,
+        id -> Nullable<Text>,
+        owner_address -> Nullable<Text>,
```
Not exactly related to this PR, but I would expect this column not to allow `NULL`? How can we have bundle without owner?

mikonieminen replied:
Should not be nullable anymore. Weren't these regenerated?

mikonieminen commented on src/database/schema.rs:
```diff
@@ -1,34 +1,34 @@
 table! {
     leaders (address) {
-        address -> Bpchar,
+        address -> Nullable<Text>,
```
Why is this changed to allow `NULL`?

mikonieminen replied:
This field is marked as `PRIMARY KEY`, how can this be nullable?

mikonieminen commented on src/database/schema.rs:
```diff
@@ -1,34 +1,34 @@
-        validated -> Bool,
-        bundle_id -> Nullable<Bpchar>,
-        sent_to_leader -> Bool,
+        id -> Nullable<Text>,
```
Why is this changed to allow `NULL`?

mikonieminen commented on src/database/schema.rs:
```diff
@@ -1,34 +1,34 @@
+        block_actual -> Nullable<Integer>,
+        signature -> Binary,
+        validated -> Integer,
+        bundle_id -> Nullable<Text>,
```
Not exactly related to this PR, but how can we have transaction without bundle ID?

bluekirby1111 replied:
Yes

It is bundled in intervals

mikonieminen replied:
OK, thanks for explaining.

mikonieminen commented on src/database/schema.rs:
```diff
@@ -1,34 +1,34 @@
     validators (address) {
-        address -> Bpchar,
-        url -> Nullable<Varchar>,
+        address -> Nullable<Text>,
```
Why is this changed to allow `NULL`?

mikonieminen replied:
This field is marked as `PRIMARY KEY`, how can this be nullable?

mikonieminen commented on src/server/routes/post_tx.rs:
```diff
@@ -41,7 +41,7 @@ pub struct ValidatorSignature {
     id: String,
     signature: String,
-    block: i64,
+    block: i32,
```
`i32` feels a bit small for block number.

mikonieminen commented on src/server/routes/sign.rs:
```diff
@@ -55,7 +55,7 @@ struct SignedBody {
 }
 
-pub async fn sign_route<Config>(
+pub async fn sign_route<'a, Config>(
```
What requires this added lifetime parameter? I cannot see any change requiring it.

mikonieminen reviewed (changes requested):
There are quite few topics that made me wonder about the changes made in this PR. One additional issue I have with this one is that it adds `unwrap`s when we should work on removing those.

randomcash committed:
Fix typing

randomcash committed:
Remove unused arg

mikonieminen commented on src/cron/arweave/arweave.rs:
```diff
@@ -108,7 +108,7 @@ pub enum ArweaveProtocol {
 pub struct Arweave {
     pub host: String,
-    pub port: i32,
+    pub port: i64,
```
I suppose changing this was an accident. Actually, I would define the port as `u16`.

mikonieminen reviewed (changes requested):
Many of my comments were marked resolved without really changing them. I find especially problematic those around database schemas that allow nullable values where to me it looks like broken data if those fields allow `NULL` values. Here is a good article explaining why: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

randomcash committed:
Add not null to fields

randomcash committed:
Fix arweave type

mikonieminen commented on src/database/models.rs:
```diff
@@ -13,7 +13,7 @@ pub struct Bundle {
 #[table_name = "bundle"]
 pub struct NewBundle {
-    pub id: String,
+    pub id: Option<String>,
```
This should not be optional after changing tables.

mikonieminen commented:
There's still something wrong around those nullable fields.

randomcash committed:
Fix schema typings

mikonieminen committed:
Make database schema closer to correct

Use `BigInt` for epoch and block number fields. While these fields,
should map to `u128`, this is step towards more correct values.

Otherwise fix fields to be `NOT NULL` where `NULL` is not accepted.
Also mark all `PRIMARY KEY` fields as `NOT NULL` so that `diesel` maps
them correctly.

mikonieminen committed:
Cleanup code to get rid of some warnings

mikonieminen committed:
Use correct type for TxReceipt::block

bluekirby1111 commented:
Sounds good to me
